crates/rust-analyzer/src/config.rs

```diff
@@ -434,6 +434,65 @@ config_data! {
+        completion_excludeFlyimportTraits: Vec<String> = Vec::new(),
+        /// A list of full paths to traits to exclude from completion.
+        ///
+        /// Methods from these traits won't be completed, even if the trait is in scope. However, they will still be suggested on expressions whose their type is `dyn Trait`, `impl Trait` or `T where T: Trait`.
+        ///
+        /// Note that the trait themselves can still be completed.
+        ///
+        /// This setting defaults to:
+        ///
+        ///  - [`core::convert::AsRef`](https://doc.rust-lang.org/nightly/core/convert/trait.AsRef.html)
+        ///  - [`core::convert::AsMut`](https://doc.rust-lang.org/nightly/core/convert/trait.AsMut.html)
+        ///  - [`core::borrow::Borrow`](https://doc.rust-lang.org/nightly/core/borrow/trait.Borrow.html)
+        ///  - [`core::borrow::BorrowMut`](https://doc.rust-lang.org/nightly/core/borrow/trait.BorrowMut.html)
+        ///  - [`core::cmp::PartialEq`](https://doc.rust-lang.org/nightly/core/cmp/trait.PartialEq.html)
+        ///  - [`core::any::Any`](https://doc.rust-lang.org/nightly/core/any/trait.Any.html)
+        ///  - All operator traits (in [`core::ops`](https://doc.rust-lang.org/nightly/core/ops))
+        ///
+        /// Note that if you override this setting, those traits won't be automatically inserted, so you may want to insert them manually.
+        completion_excludeTraits: Vec<String> = vec![
+            "core::convert::AsRef".to_owned(),
+            "core::convert::AsMut".to_owned(),
+            "core::borrow::Borrow".to_owned(),
+            "core::borrow::BorrowMut".to_owned(),
+            "core::cmp::PartialEq".to_owned(),
+            "core::any::Any".to_owned(),
+            "core::ops::Add".to_owned(),
+            "core::ops::AddAssign".to_owned(),
+            "core::ops::BitAnd".to_owned(),
+            "core::ops::BitAndAssign".to_owned(),
+            "core::ops::BitOr".to_owned(),
+            "core::ops::BitOrAssign".to_owned(),
+            "core::ops::BitXor".to_owned(),
+            "core::ops::BitXorAssign".to_owned(),
+            "core::ops::Div".to_owned(),
+            "core::ops::DivAssign".to_owned(),
+            "core::ops::Mul".to_owned(),
+            "core::ops::MulAssign".to_owned(),
+            "core::ops::Rem".to_owned(),
+            "core::ops::RemAssign".to_owned(),
+            "core::ops::Shl".to_owned(),
+            "core::ops::ShlAssign".to_owned(),
+            "core::ops::Shr".to_owned(),
+            "core::ops::ShrAssign".to_owned(),
+            "core::ops::Sub".to_owned(),
+            "core::ops::SubAssign".to_owned(),
+            "core::ops::Neg".to_owned(),
+            "core::ops::Not".to_owned(),
+            "core::ops::Index".to_owned(),
+            "core::ops::IndexMut".to_owned(),
+            "core::ops::Deref".to_owned(),
+            "core::ops::DerefMut".to_owned(),
+        ],
```

Comment:
I don't like these defaults :)

Imo, if an operator is in fact in scope, its methods should be completed. 

Also `AsRef/AsMut` should definitely not be in here, these are frequently used (they are in the prelude for a reason).

In my opinion, this list should be empty and the two Borrow traits should be put into the above config.

Reply:
> Imo, if an operator is in fact in scope, its methods should be completed.

Do you want to move them the other config then? Or remove them completely?

> Also `AsRef/AsMut` should definitely not be in here, these are frequently used (they are in the prelude for a reason).

At least in std (but not only), all `AsRef/Mut` implementations have an alternative, that is better because it does not require type ascription and is more clear. So aside from generic code, I can't understand why people would want that. Still, I'm not going to argue about the defaults, so if you feel this is better I can do that.

> In my opinion, this list should be empty and the two Borrow traits should be put into the above config.

What with `Any`?

Reply:
> Do you want to move them the other config then? Or remove them completely?

Tough to say, I personally do import `ops::Not` by writing `.not` from time to time, but none of the others. So moving them to the other config might be fine?

> At least in std (but not only), all `AsRef/Mut` implementations have an alternative, that is better because it does not require type ascription and is more clear. So aside from generic code, I can't understand why people would want that. Still, I'm not going to argue about the defaults, so if you feel this is better I can do that.

I don't think that's true? I very frequently use `str`'s `AsRef<Path>` implementation for example, there is no alternative method on `str` for this. I don't consider using `Path::new` instead an alternative as that requires me to import `Path` unnecessarily when the output of the `as_ref` call can be inferred from the call site. We also use this pattern a lot for our `AbsPath` (and similar) wrappers.
 
> What with `Any`?

I didn't even realize that was in the list. I don't think we need to add this to either config, `type_id` is a fairly long and unique method name (unlike operator methods and the rest).

Reply:
At the very least put this list in the documentation as a "you probably want to look into excluding these" list

Reply:
> At the very least put this list in the documentation as a "you probably want to look into excluding these" list

I don't think that's needed, and I also don't really find a place for that.

Reply:
I am also unresolving this, I still think the defaults aren't great. They should only list the `Borrow` and `BorrowMut` traits in my eyes.